schema.py

```diff
@@ -117,6 +124,8 @@ def validate(self, data):
             coverage = set()  # matched schema keys
             # for each key and value find a schema entry matching them, if any
             sorted_skeys = sorted(s, key=priority)
+            # resort sorted_skeys with priority_of_dict_key
+            sorted_skeys = sorted(sorted_skeys, key=priority_of_dict_key)
```

Comment:
Sorry, I believe I misunderstood what you're trying to achieve.

You're aiming for an order like this one, is that right?
```python
'x', Optional('y'), int, Optional(str)
```

If that's right, I'd propose that we have a static method in `Schema` roughly like this:

```python
@staticmethod
def _dict_key_priority(key):
    if isinstance(key, Optional):
        return priority(key._schema) + 0.5
    return priority(key)
```

That's still hacky but hopefully a bit clearer than sorting twice.